### ISSUE TITLE
Issue between FileHandler and AdminWebServer causes the whole file to be read in memory by AWS

### ISSUE THREAD

**Issue by wil93** [cms-dev/cms#870]
To reproduce, import a task with a single 1GB testcase and then try to download said testcase from AWS.

With @bblackham I investigated this a bit, we tried using `.flush()` after `.write()` but it wasn't enough. We tried adding `Transfer-Encoding: chunked` but it appears it's better to not use that (it's automatically added)

**Comment by lerks** [comment]
Do you have more details? You're mentioning that it's an issue between FileHandler and AWS: do you know what the issue is? what code is causing it? I looked at FileHandler and apart from the missing `flush` everything seems fine.

**Comment by wil93** [comment]
I just see the AdminWebServer process taking up my RAM while my browser has not even started downloading anything and, when AdminWebServer finished taking up my RAM, the download finally starts.

We tried adding flush but AdminWebServer was still taking up RAM (without starting the download) in the same way

**Comment by wil93** [comment]
This is the patch I applied to test `flush()`

```patch
diff --git a/cms/server/util.py b/cms/server/util.py
index b4ea4cc0..0a175478 100644
--- a/cms/server/util.py
+++ b/cms/server/util.py
@@ -335,6 +335,7 @@ def file_handler_gen(BaseClass):
             ret = True
             while ret:
                 ret = self._fetch_write_chunk()
+                logger.warning("I sent a chunk!!")
                 gevent.sleep(0)
 
         def _fetch_write_chunk(self):
@@ -343,13 +344,14 @@ def file_handler_gen(BaseClass):
             """
             data = self.temp_file.read(FileCacher.CHUNK_SIZE)
             length = len(data)
-            self.size += length / (1024 * 1024)
+            self.size += length
             self.write(data)
+            self.flush()
             if length < FileCacher.CHUNK_SIZE:
                 self.temp_file.close()
                 duration = time.time() - self.start_time
                 logger.info("%.3lf seconds for %.3lf MB",
-                            duration, self.size)
+                            duration, self.size / (1024 * 1024))
                 self.finish()
                 return False
             return True
```

**Comment by lerks** [comment]
I think I debugged this. Calling `self.flush()` causes the Tornado handler to call the `write` method on the connection. In our case, this is a `_WSGIConnection`, and its write method simply adds the chunk to a buffer:
https://github.com/tornadoweb/tornado/blob/3fc1f622a1ac4d098f260d1efe4162ca04eeabf7/tornado/wsgi.py#L128

The buffer is only "emptied" at the very end of the request, just before the WSGI handler returns:
https://github.com/tornadoweb/tornado/blob/3fc1f622a1ac4d098f260d1efe4162ca04eeabf7/tornado/wsgi.py#L227

I'm not really sure what we can do about it...

**Comment by wil93** [comment]
Is it worth reporting to tornado?

**Comment by wil93** [comment]
FWIF: I didn't investigate yet if this causes the web server to load in RAM a copy of the file **for each request** (I would say yes, but I'm not sure), but if this is the case then we have a possible denial of service on our hands: could a contestant make thousands of requests PDF statement to ContestWebServer and drain the server's memory?

**Comment by lerks** [comment]
I opened an issue on Tornado about this: https://github.com/tornadoweb/tornado/issues/2294

I'm not expecting much from it, though: as I mentioned recently on Gitter their WSGI support is hacky and sub-standard almost by design.

And, yes, I think what you said in your last comment is correct. Though files downloadable through CWS are usually very small, and thus the ability to make use of this bug should be limited.

I'm considering possible fixes for this, the main one consisting in implementing our own WSGI application for serving files, rather than building it on top of a Tornado handler.

**Comment by wil93** [comment]
> Though files downloadable through CWS are usually very small, and thus the ability to make use of this bug should be limited

PDFs can reach a couple MBs sometimes (meaning that a user can just make a couple thousand requests and shut down a server with low RAM available)

Although the biggest issue is probably with OutputOnly tasks: the input data can be very big sometimes.

**Comment by lerks** [comment]
@wil93 could you cherry-pick these two commits (https://github.com/lerks/cms/commit/7e47b4d25af63f378584a806292ce36021a5da86 https://github.com/lerks/cms/commit/2fe7cae966ca355fb2fefdf0296684e6e5098abc)  and let me know if you still experience this issue?

**Comment by wil93** [comment]
Done. The two commits appear to solve the issue completely :+1: :+1: :+1:

**Comment by lerks** [comment]
Fixed in #984.
